circleguard/widgets.py

```diff
@@ -325,138 +334,372 @@ def __init__(self, tooltip):
-        layout.addWidget(label, 0, 0, 1, 1)
-        layout.addItem(SPACER, 0, 1, 1, 1)
-        layout.addWidget(self.combined, 0, 2, 1, 3)
+        super().__init__("Map Replay", [self.map_id_input, self.user_id_input])
+
+        self.layout.addWidget(self.map_id_input, 1, 0, 1, 8)
+        self.layout.addWidget(self.user_id_input, 2, 0, 1, 8)
+        self.layout.addWidget(self.mods_input, 3, 0, 1, 8)
 
-        self.setLayout(layout)
 
-    # keep spinbox and slider in sync
-    def update_spinbox(self, value):
-        self.spinbox.setValue(value)
+class ReplayPathW(LoadableW):
+    def __init__(self):
+        self.path_input = FolderChooser(".osr path", folder_mode=False, file_ending="osu! Replayfile (*.osr)")
+
+        super().__init__("Local Replay", [self.path_input])
+
+        self.layout.addWidget(self.path_input, 1, 0, 1, 8)
+
+    def check_required_fields(self):
+        for input_widget in self.required_input_widgets:
+            all_filled = True
+            filled = input_widget.changed
+            if not filled:
+                input_widget.show_required()
+                all_filled = False
+        return all_filled
+
+class MapW(LoadableW):
+    def __init__(self):
+
+        self.map_id_input = InputWidget("Map id", "", "id")
+        self.span_input = InputWidget("Span", "", "normal")
+        self.span_input.field.setPlaceholderText("1-50")
+        self.mods_input = InputWidget("Mods (opt.)", "", "normal")
+
+        super().__init__("Map", [self.map_id_input, self.span_input])
+
+        self.layout.addWidget(self.map_id_input, 1, 0, 1, 8)
+        self.layout.addWidget(self.span_input, 2, 0, 1, 8)
+        self.layout.addWidget(self.mods_input, 3, 0, 1, 8)
 
-    def update_slider(self, value):
-        self.slider.setValue(value)
+    def check_required_fields(self):
+        for input_widget in self.required_input_widgets:
+            all_filled = True
+            # don't count span_input as empty when it has placeholder text
+            filled = input_widget.field.text() != "" or input_widget.field.placeholderText() != ""
+            if not filled:
+                input_widget.show_required()
+                all_filled = False
+        return all_filled
+
+class UserW(LoadableW):
+    def __init__(self):
+
+        self.user_id_input = InputWidget("User id", "", "id")
+        self.span_input = InputWidget("Span", "", "normal")
+        self.mods_input = InputWidget("Mods (opt.)", "", "normal")
+
+        super().__init__("User", [self.user_id_input, self.span_input])
+
+        self.layout.addWidget(self.user_id_input, 1, 0, 1, 8)
+        self.layout.addWidget(self.span_input, 2, 0, 1, 8)
+        self.layout.addWidget(self.mods_input, 3, 0, 1, 8)
+
+
+class MapUserW(LoadableW):
+    def __init__(self):
+        self.map_id_input = InputWidget("Map id", "", "id")
+        self.user_id_input = InputWidget("User id", "", "id")
+        self.span_input = InputWidget("Span", "", "normal")
```

Comment:
I suggest a default of 1-3. (or make it optional)

Reply:
if it were optional it would load all of them? If so I think it would be better to be explicit about it and fill it with an `"all"` placeholder text, since we don't have optional spans anywhere else, I think that would confuse people

Reply:
ye "all" is fine imo